### ISSUE TITLE
Upgrade Trusty to Xenial

### ISSUE THREAD

**Issue by fenris** [Microsoft/BashOnWindows#482]
I've tried to upgrade from Trusty to Xenial , result ? crash due to missing path and error writing files installations.

Is there inside the milestone to place xenial based bash in future Windows Insiders Fast Ring update?

**Comment by fpqc** [comment]
fenris, Canonical's new image for WSL should be up on the Windows store I think with the next insider update.

**Comment by sunilmut** [comment]
WSL currently does not support Xenial officially and that's not on plan for the anniversary update. But, that is definitely something we will be looking to support in (during) the next release.

**Comment by paladox** [comment]
@fpqc Hi the next insider release has been released is the image on the windows store yet.

**Comment by sunilmut** [comment]
@fqpc - That is correct, Windows build 14361 has been released to insider preview. But just FYI that the Windows build and the Ubuntu image that comes from the Windows store and not tied together. The are and can be updated independently.

**Comment by paladox** [comment]
@sunilmut hi is @fpqc correct about the image being updated on the windows store please.

**Comment by sunilmut** [comment]
@paladox - This is on our radar, but we don't have any ETA on when the Xenial image will be uploaded on the Windows store.

**Comment by paladox** [comment]
Ok thanks for replying.

**Comment by prwang** [comment]
You may simply unroll the squashfs file of edition Xenial and chroot into it, and you may use Archlinux or Fedora the same way.

**Comment by Manouchehri** [comment]
I have Xenial (16.04) working in WSL.

![Bash on Ubuntu 16.04 on Windows](https://cloud.githubusercontent.com/assets/7232674/16308939/f0858428-3934-11e6-9760-77878852a165.png)

**Comment by benhillis** [comment]
@Manouchehri Very nice!  Any special workarounds you needed to get this working?  I'd be very interested in hearing the process.  I'm sure my friend @dustinkirkland from Canonical would be interested too :)

**Comment by paladox** [comment]
@Manouchehri how did you get it installed please. And did you have to clean install or did you manage to get your files upgraded without loosing any data.

**Comment by Manouchehri** [comment]
Worked on my first try, nothing too special.

```bash
sudo do-release-upgrade -f DistUpgradeViewNonInteractive -d # Screen is broken.
```

It still gets stuck at a yes/no type prompt, so I ctrl+c'd out. To fix the remaining packages, just run dpkg.

```bash
 sudo dpkg --configure -a
```

After that I ran `apt-get update`, `apt-get upgrade`, `apt-get dist-upgrade` and `apt-get autoremove`.

**Comment by paladox** [comment]
@Manouchehri thanks, yay it works.

**Comment by paladox** [comment]
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04 LTS
Release:        16.04
Codename:       xenial

:)

**Comment by dustinkirkland** [comment]
+100 xp :-)


:-Dustin

**Comment by paladox** [comment]
@dustinkirkland and @benhillis I get this error

snap find
2016/06/23 20:30:05.121452 main.go:156: WARNING: can not create syslog logger
error: cannot list snaps: cannot communicate with server: Get http://localhost/v2/find?q=: dial unix /run/snapd.socket: setsockopt: invalid argument

when trying to run snap.

**Comment by Manouchehri** [comment]
That's snapd trying to use systemd's logger. Previously there was separate logging daemon, so init (Uptrack) didn't cause an issue. (There's no systemd or Uptrack support in Ubuntu on Windows from what I understand.)

To fix it, we could probably bring back syslog-ng until systemd works. This is probably @dustinkirkland's job. ;)

**Comment by benhillis** [comment]
@paladox This particular issue looks like missing SO_BROADCAST flag support for Unix sockets.  #568 is tracking the same issue.  I suspect snap will hit other unimplemented things as well as it's a binary that we haven't tried much.

**Comment by Manouchehri** [comment]
Oops, I stopped reading after the logging warning line.. What @benhillis said sounds correct as for the socket error. =)

**Comment by paladox** [comment]
Oh, thanks for replying.

**Comment by Manouchehri** [comment]
@paladox Does it actually work on 14.04 either?

**Comment by benhillis** [comment]
@Manouchehri you're probably right also :)

**Comment by paladox** [comment]
@Manouchehri I'm not sure. I didn't hit that problem before. But I doint think snaps are on Ubuntu 14.04.

**Comment by Manouchehri** [comment]
I found another fun error. Probably PEBKAC related.

```bash
ubuntu@DESKTOP-3RQO5S5:~$ ps aux
Error: /proc must be mounted
  To mount /proc at boot you need an /etc/fstab line like:
      proc   /proc   proc    defaults
  In the meantime, run "mount proc /proc -t proc"
ubuntu@DESKTOP-3RQO5S5:~$ mount
rootfs on / type rootfs (rw,relatime)
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000)
tmpfs on /run type tmpfs (rw,nosuid,noexec,relatime,size=204320k,mode=755)
none on /run/lock type tmpfs (rw,nosuid,nodev,noexec,relatime,size=5120k)
none on /run/shm type tmpfs (rw,nosuid,nodev,relatime)
none on /run/user type tmpfs (rw,nosuid,nodev,noexec,relatime,size=102400k,mode=755)
```

**Comment by benhillis** [comment]
@Manouchehri I took a look at what's causing ps to not work in 16.04.  As funny as it sounds, it is due to us missing the `/proc/sys/vm/min_free_kbytes` ProcFs file.  I just wrote a change to stubbed out that file and ps is mostly working.  TTY is showing up as a '?' which is caused by us not having a /proc/tty directory.

**Comment by dustinkirkland** [comment]
Snaps are new, as of 16.04, and generally required systemd to work.


:-Dustin

**Comment by paladox** [comment]
@benhillis is there anyway we could support systemd  please and also support /proc/tty please.

**Comment by Manouchehri** [comment]
@benhillis Ha, I just noticed that too.

```bash
read(4, "MemTotal:        8313236 kB\nMemF"..., 8191) = 1164
open("/proc/sys/vm/min_free_kbytes", O_RDONLY) = -1 ENOENT (No such file or directory)
write(2, "Error: /proc must be mounted\n  T"..., 176Error: /proc must be mounted
```

If I want to screw around in `/proc/sys/vm/`, what's the proper way to do that here?

```bash
root@DESKTOP-3RQO5S5:~# strace touch /proc/sys/vm/min_free_kbytes
execve("/usr/bin/touch", ["touch", "/proc/sys/vm/min_free_kbytes"], [/* 19 vars */]) = 0
brk(NULL)                               = 0x1a99000
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fca1ed90000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=26136, ...}) = 0
mmap(NULL, 26136, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fca1ed97000
close(3)                                = 0
access("/etc/ld.so.nohwcap", F_OK)      = -1 ENOENT (No such file or directory)
open("/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0P\t\2\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0755, st_size=1864888, ...}) = 0
mmap(NULL, 3967488, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7fca1e630000
mprotect(0x7fca1e7f0000, 2093056, PROT_NONE) = 0
mmap(0x7fca1e9ef000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1bf000) = 0x7fca1e9ef000
mmap(0x7fca1e9f5000, 14848, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7fca1e9f5000
close(3)                                = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fca1ed80000
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fca1ed70000
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fca1ed60000
arch_prctl(ARCH_SET_FS, 0x7fca1ed70700) = 0
mprotect(0x7fca1e9ef000, 16384, PROT_READ) = 0
mprotect(0x60e000, 4096, PROT_READ)     = 0
mprotect(0x7fca1ec25000, 4096, PROT_READ) = 0
munmap(0x7fca1ed97000, 26136)           = 0
brk(NULL)                               = 0x1a99000
brk(0x1aba000)                          = 0x1aba000
open("/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=2981280, ...}) = 0
mmap(NULL, 2981280, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fca1e358000
close(3)                                = 0
open("/proc/sys/vm/min_free_kbytes", O_WRONLY|O_CREAT|O_NOCTTY|O_NONBLOCK, 0666) = -1 EPERM (Operation not permitted)
utimensat(AT_FDCWD, "/proc/sys/vm/min_free_kbytes", NULL, 0) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/locale.alias", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=2995, ...}) = 0
read(3, "# Locale name alias data base.\n#"..., 512) = 512
read(3, "See the\n# GNU General Public Lic"..., 512) = 512
read(3, "nd for the time being for\n# back"..., 512) = 512
read(3, "hese lines were removed\n# becaus"..., 512) = 512
read(3, "8859-1\ngalego\t\tgl_ES.ISO-8859-1\n"..., 512) = 512
read(3, "O.ISO-8859-1 nb_NO.ISO-8859-1\nno"..., 512) = 435
read(3, "", 512)                        = 0
close(3)                                = 0
open("/usr/share/locale/en_US.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en_US.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en_US/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en_US.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en_US.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en_US/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en.UTF-8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en.utf8/LC_MESSAGES/coreutils.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en/LC_MESSAGES/coreutils.mo", O_RDONLY) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=619, ...}) = 0
mmap(NULL, 619, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7fca1ed9d000
close(3)                                = 0
write(2, "touch: ", 7touch: )                  = 7
write(2, "setting times of '/proc/sys/vm/m"..., 47setting times of '/proc/sys/vm/min_free_kbytes') = 47
open("/usr/share/locale/en_US.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en_US.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en_US/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale/en/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en_US.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en_US.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en_US/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en.utf8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
open("/usr/share/locale-langpack/en/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
write(2, ": No such file or directory", 27: No such file or directory) = 27
write(2, "\n", 1
)                       = 1
close(1)                                = 0
close(2)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by Manouchehri** [comment]
@benhillis Somewhat off topic, but can I get debugging symbols for the Insider Preview builds? Shoot me an email if you'd prefer to talk about that privately.

**Comment by benhillis** [comment]
@Manouchehri The ProcFs filesystem is a virtual file system exposed by our kernel drivers (lxcore.sys and lxss.sys) so unfortunately you can't change them without building a new driver.

Regarding debugging symbols: I know we release public versions of the symbols, I'm not sure about the process for Insider Preview builds though. @russalex would know the right answer to this question.  Have you tried [Using the Microsoft Symbol Server](https://msdn.microsoft.com/en-us/library/windows/desktop/ee416588(v=vs.85).aspx#using_the_microsoft_symbol_server)?

**Comment by fenris** [comment]
fenris@X240:~$ cat /etc/issue
Ubuntu 16.04 LTS \n \l
fenris@X240:~$ sudo apt-get upgrade
sudo: no tty present and no askpass program specified
fenris@X240:~$ ps aux
Error: /proc must be mounted
  To mount /proc at boot you need an /etc/fstab line like:
      proc   /proc   proc    defaults
  In the meantime, run "mount proc /proc -t proc"
fenris@X240:~$

**Comment by Manouchehri** [comment]
@fenris The fix hasn't been pushed yet, give it a few days.

@benhillis The public symbol server doesn't seem to have any of the GUIDs from Insider builds.

```cmd
0:006> .reload /f
Reloading current modules
.*** ERROR: Module load completed but symbols could not be loaded for bash.exe
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\Windows\System32\msvcp_win.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\Windows\System32\lxss\LxssManagerProxyStub.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\kernel.appcore.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\bcryptPrimitives.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\ucrtbase.dll - 


Press ctrl-c (cdb, kd, ntsd) or ctrl-break (windbg) to abort symbol loads that take too long.
Run !sym noisy before .reload to track down problems loading symbols.

.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\KERNELBASE.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\RPCRT4.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\combase.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\msvcrt.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\KERNEL32.DLL - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\sechost.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\WINDOWS\System32\clbcatq.dll - 
.*** ERROR: Symbol file could not be found.  Defaulted to export symbols for ntdll.dll - 


************* Symbol Loading Error Summary **************
Module name            Error
bash                   The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
msvcp_win              The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
LxssManagerProxyStub   The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
kernel.appcore         The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
bcryptPrimitives       The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
ucrtbase               The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
KERNELBASE             The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
RPCRT4                 The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
combase                The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
msvcrt                 The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
KERNEL32               The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
sechost                The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
clbcatq                The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols
ntdll                  The system cannot find the file specified : srv*c:\symbols*https://msdl.microsoft.com/download/symbols

You can troubleshoot most symbol related issues by turning on symbol loading diagnostics (!sym noisy) and repeating the command that caused symbols to be loaded.
You should also verify that your symbol search path (.sympath) is correct.
```

**Comment by Manouchehri** [comment]
I also checked for symbols on bash.exe from 10.0.14316.1000, no luck with that either.

![image](https://cloud.githubusercontent.com/assets/7232674/16321379/7e79906e-396a-11e6-9a40-8f62c7f9bda0.png)

**Comment by aschneiderg** [comment]
@fenris 

<code>echo 'yourpasswordhere' | sudo -S dpkg --configure -a</code>

**Comment by fenris** [comment]
@alvaro thanks its help

but after finished everything., I close the bash and reopened back ...  its
not permanent.

Do I need to run echo and dpkg --configure everythime I open bash?

**Comment by therealkenc** [comment]
I've been using Xenial since Saturday -- I was inspired by @aseering over in the #543 thread. The method I used we just to copy my /etc/apt directory over from a native Ubuntu 16.04 and did `apt update`. Hit the same problems with `sudo` and `procps`. `strace` also had issues. For now I just worked around by rebuilding those from upstream (unpatched) 14.04 sources and installed in /usr/local.

**Comment by benhillis** [comment]
@fenris after running the do-release-upgrade command did you run 'sudo dpkg --configure -a'

**Comment by fenris** [comment]
@benhillis yes I did ... it shown nothing ...

**Comment by hazzino** [comment]
Best thing to get an upgrade is to remember to give a root password a part the user one during the installation time.It'd would be nice if during the final stage the grub stuff will be closed.
I hope a light in the near next about docker stuff and raw socket

**Comment by mphaney** [comment]
Well, shoot.  I can't even seem to get past the first command :-/  Oh well.  This is still cool though!

```
$ sudo do-release-upgrade -f DistUpgradeViewNonInteractive -d
Checking for a new Ubuntu release
Get:1 Upgrade tool signature [198 B]
Get:2 Upgrade tool [1,265 kB]
Fetched 1,265 kB in 0s (0 B/s)
authenticate 'xenial.tar.gz' against 'xenial.tar.gz.gpg'
extracting 'xenial.tar.gz'
pcilib: Cannot open /proc/bus/pci
lspci: Cannot find any working access method.
```

**Comment by Manouchehri** [comment]
Did you let it continue to run? Errors are normal and can safely be
ignored.

**Comment by mphaney** [comment]
Sorry, I failed to copy the last line (I'll update my comment).  It returned to a prompt immediately after.  Unless it runs in the background, it appeared to have failed completely.

**Comment by Manouchehri** [comment]
Try running `sudo dpkg --configure -a` to continue.

**Comment by mphaney** [comment]
Ah, interesting.  The first time I ran `sudo dpkg --configure -a` it simply returns to the prompt without doing anything.  HOWEVER, I went ahead and proceeded to run `apt-get update`, `apt-get upgrade`, `apt-get dist-upgrade` and finally `apt-get autoremove`, THEN went back and ran the original

`sudo do-release-upgrade -f DistUpgradeViewNonInteractive -d`

Now it is proceeding with the upgrade.  Guess I'll let it follow through and see what happens :-)

**Comment by mphaney** [comment]
Ok, so here's the final state of things on my end.  From a brand new install of Bash on build 14379, ran the following:

```
sudo su
apt-get update
apt-get upgrade
apt-get dist-upgrade
apt-get autoremove
sudo do-release-upgrade -f DistUpgradeViewNonInteractive -d
```

As noted, it still gets stuck at a Y/N prompt.  Pressed Ctrl+C and ran:

```
echo 'yourpasswordhere' | sudo -S dpkg --configure -a
apt-get update
apt-get upgrade
apt-get dist-upgrade
apt-get autoremove
```

And when all is said and done, I see this:

```
$ cat /etc/issue
Ubuntu 16.04 LTS \n \l
```
Sweet :-)

**Comment by fpqc** [comment]
@therealkenc How did you rebuild those utilities from source and then install to /usr/local without patching?  Won't they install to / by default?

**Comment by SandHilt** [comment]
After update i have issues with sudo in default user. How i fix?
```bash
sudo: no tty present and no askpass program specified
```

**Comment by fpqc** [comment]
@sandhilt KenC said he recompiled sudo from trusty.  That's actually exactly what I was asking =).

**Comment by therealkenc** [comment]
I used procps-ng-3.3.9. Add `--without-ncurses` to `./configure`.

**Comment by wakefu1** [comment]
To make sudo work just use sudo -S <command>

**Comment by jamespacileo** [comment]
Am I wrong or there's no `/mnt/c/` ? :smile:

**Comment by Manouchehri** [comment]
There should be?

**Comment by jamespacileo** [comment]
@Manouchehri tried wiping and running it again. 

The session where the upgrade was made if done in /mnt/, the folder structure/content stay visible e.g. `ls` works

`/mnt/` is currently empty anyway. Is your `/mnt/c/...` there?

**Comment by jamespacileo** [comment]
Ok for anyone who encountered my same problems:

- If `/mnt/c/` disappears after upgrade, reboot Windows.

- if you get the following error `sudo: no tty present and no askpass program specified`, you can temporarily change main user to root with `lxrun /setdefaultuser root`

**Comment by fpqc** [comment]
@jamespacileo Yeah, once you do that, I suggest following my instructions and @therealkenc 's to compile and install (to /usr/local) Ubuntu Trusty's versions of sudo, strace, and procps from the Trusty source repo.  This should let you use Xenial in WSL as a non-root user again.

**Comment by therealkenc** [comment]
FWIW, I am pretty sure I went about it the hard way. Should be able to do something like the following. 

From cmd:
`lxrun /setdefaultuser root`

From bash as root: 
```
wget http://mirrors.kernel.org/ubuntu/pool/main/s/sudo/sudo_1.8.9p5-1ubuntu1.1_amd64.deb
wget http://mirrors.kernel.org/ubuntu/pool/main/p/procps/procps_3.3.9-1ubuntu2_amd64.deb
wget http://mirrors.kernel.org/ubuntu/pool/main/s/strace/strace_4.8-1ubuntu5_amd64.deb
dpkg -i sudo_1.8.9p5-1ubuntu1.1_amd64.deb
dpkg -i procps_3.3.9-1ubuntu2_amd64.deb
dpkg -i strace_4.8-1ubuntu5_amd64.deb
```

Haven't tried mind you.

**Comment by onomatopellan** [comment]
@therealkenc It works. Just finished upgrading to Xenial and your method fixed the sudo, ps problems. Thanks!

**Comment by SandHilt** [comment]
@onomatopellan when i update my packages, the newest version appear...
So you can try this workaround:
```bash
echo "alias sudo='sudo -S'" >> ~/.bash_aliases
```
Because you redirect to 'askpass' to terminal.

**Comment by SandHilt** [comment]
Or, if you have problems with newest packages, you can keep versions. Try this, after install packages @therealkenc :
```bash
sudo -S apt-mark hold sudo
sudo -S apt-mark hold procps
sudo -S apt-mark hold strace
```
source: [here](http://askubuntu.com/questions/18654/how-to-prevent-updating-of-a-specific-package)

**Comment by denyardy** [comment]
after upgrade 16.04, i can't use ping

`root@JNCK-PC:/home/denyardy# ping google.com
ping: icmp open socket: Permission denied`

**Comment by onomatopellan** [comment]
@denyardy You need to open bash with "Run as Administrator" if you want to use ping.

**Comment by denyardy** [comment]
@onomatopellan thanks

**Comment by onomatopellan** [comment]
Just upgraded to Xenial on build 14901. This time there were no hangs but I had to `sudo apt-get dist-upgrade -f` a couple of times to fix some dependences problems. After that sudo, ps, strace work well without needing old packages.

**Comment by fpqc** [comment]
@onomatopellan That's amazing!

**Comment by yurikoles** [comment]
````
$ cat /etc/issue
Debian GNU/Linux stretch/sid \n \l
```

**Comment by therealkenc** [comment]
Good place to post this as any... I upgraded WSL to Yakety today, partly inspired by @fpqc's [comment](https://github.com/Microsoft/BashOnWindows/issues/637#issuecomment-242951010) in #637. Went much smoother than Xenial back in June.

Hit a problem I didn't expect though: `locale-gen` fails at `clone` syscall:
```
clone(child_stack=0x7ff6e6340ff0, flags=CLONE_VM|CLONE_VFORK|SIGCHLD) 
    = -1 EINVAL (Invalid argument)
```
Unfortunately `locale`/`localedef`/`locale-gen` are buried in the glibc package so I haven't tracked it down enough to know what exactly is going on there. Closest I could find is [here](https://github.com/bminor/glibc/blame/2a69f853c03034c2e383e0f9c35b5402ce8b5473/sysdeps/unix/sysv/linux/ia64/vfork.S#L30), but I have a tough time imagining [vfork](http://linux.die.net/man/2/vfork) not working all this time. 

Downgrading isn't an option because you'd have to take all of glibc with you. As a work around I just copied `/usr/share/locale` from native Yakety. YRMV.

**Comment by fpqc** [comment]
@therealkenc I already figured it out.  You need to un-gzip the locale file.  If you used that python installer I recommended with the sample script, it does it for you.

**Comment by therealkenc** [comment]
Yes unzipping it is the same as copying over. It's the failing clone that's curious. Unless `sudo locale-gen` works for you? Been broken forever and I've just never noticed maybe?

**Comment by fpqc** [comment]
@therealkenc I dunno, maybe Microsoft was just ungzipping them on setup instead of running locale-gen?

**Comment by therealkenc** [comment]
Nah, wrong layer. MS doesn't mess with the Canonical packages (and logistically speaking, _can't_ mess with them). _Something_ changed. There's a [changelog](https://raw.githubusercontent.com/bminor/glibc/master/ChangeLog) entry related to `CLONE_VM` on 2016-04-29, but I would think any change that actually broke `vfork` would take down the whole house, not  just `locale-gen`.

**Comment by fpqc** [comment]
I have no idea, seems weird.  All the more reason to file it as its own issue, I think.  

Can you reproduce it on a clean install of yakkety?

**Comment by therealkenc** [comment]
I did the upgrade by changing `sources.list`, like I did with Xenial back in June. It's clean as far as it goes. Right I don't have a well formed test case that doesn't open with "Install an unsupported version of Ubuntu and..." so I'll wait on the bug report for now. Be different if I took the trouble to pin a test case.

**Comment by therealkenc** [comment]
Confirmed it's in the bones. I just got the same error from `clang` with glibc 2.4:

`clang: error: unable to execute command: posix_spawn failed: Operation not permitted`

`strace` shows the same `CLONE_VM|CLONE_VFORK|SIGCHLD` construct. Maybe I'll put together a test case on a higher motivation day. It won't exist in the wild on Trusty or Xenial because the problem appears to be down in glibc. That is, there's nothing wrong with `clang` or `locale-gen`. Or at least that's how it looks to me right now.

**Comment by fpqc** [comment]
@therealkenc Yeah, I figured that it would have to be glibc, because that's where you'd see a change in ABI on the usermode side.

**Comment by lapo-luchini** [comment]
After upgrade to Xenial `sudo` stopped being able to find a pty and ask for a password; I worked that around using `NOPASSWD`.

**Comment by therealkenc** [comment]
The `vfork` API didn't change; it _looks_ like glibc might be doing something new internally that WSL does not support. The `clone(2)` syscall API hasn't changed since Linux 2.6.49 or thereabouts. I won't know the details until/unless I put together a test case.

**Comment by fpqc** [comment]
@therealkenc Oh, that's good to know.  I thought this might be an example of some kind of breaking change in the Syscall interface.

**Comment by dorcas-ay** [comment]
Is `gnome-terminal` still working for anybody in Xenial? 
`$ gnome-terminal`

`libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
Error constructing proxy for org.gnome.Terminal:/org/gnome/Terminal/Factory0: Exhausted all available authentication mechanisms (tried: EXTERNAL, DBUS_COOKIE_SHA1, ANONYMOUS) (available: EXTERNAL, DBUS_COOKIE_SHA1, ANONYMOUS)`

**Comment by fpqc** [comment]
@dorcas-ay You might need to redo the dbus trick.

**Comment by dorcas-ay** [comment]
Thanks @fpqc. I did but it's still not working

**Comment by dorcas-ay** [comment]
I got it back running by disabling external authentication in the 'session.conf' file.

**Comment by MrObvious** [comment]
What exactly did you do?

**Comment by fpqc** [comment]
@russalex Hey, since we've done a lot of testing now with Xenial, and there are no more sudo and ps bugs, why not switch the lxrun installer to default to the Xenial image in 14915 or the next insider build?  Since it looks like Xenial performs as well as Trusty now, wouldn't you prefer to support/test against a more up to date configuration?

**Comment by dorcas-ay** [comment]
@MrObvious I commented out this line in '/etc/dbus-1/session.conf' file.

So it look like this.

'#<auth>EXTERNAL</auth>'

**Comment by russalex** [comment]
@fpqc We do have plans on moving over to Xenial / making Xenial available as the install.  I do believe it's the right way to go.  I am also grateful for people running / testing on it (and run it myself on my primary machine).  Completely agree that we need to have people testing on Xenial soon.

There are a few things we want to do before making Xenial available for direct install.  Until then, we're looking to squash the bugs that make the upgrade process painful.  An example, the main reason screen was high on our priority list was to support the upgrade process.

We'll have more progress / more to say here as the insider builds continue.

**Comment by iz0eyj** [comment]
@russalex  I'm using heavily Xenial from a couple of days and it looks really solid, but is correct first resolve the current problems.
Can I ask if there is an ETA for dbus?

**Comment by fpqc** [comment]
@iz0eyj It seems to be working now, although there's a weird authentication thing now with some things on dbus.

**Comment by iz0eyj** [comment]
@fpqc Hopefully it won't take too long, many users would like to try the WSL (chart) but the problem almost trivial of dbus becomes very complicated for someone who has never used a Terminal.
Also here I see many users participating in the beta testing with its non-Insider version, and that in the long could cause a mess.

**Comment by MrObvious** [comment]
@dorcas-ay 
> @MrObvious I commented out this line in /etc/dbus-1/session.conf file.
> So it look like this.
>#<auth>EXTERNAL</auth>

I don't think it's working. 
ERROR 2016-09-05 09:26:36 unity.glib.dbus.proxy GLibDBusProxy.cpp:203 Unable to connect to proxy: Exhausted all available authentication mechanisms (tried: EXTERNAL, DBUS_COOKIE_SHA1, ANONYMOUS) (available: EXTERNAL, DBUS_COOKIE_SHA1, ANONYMOUS)


Also @russalex is there a way to push fixes out sooner? What'd be nice is if it could be done out of band of windows updates for full blown releases so we don't have to pull down a new image.

**Comment by dorcas-ay** [comment]
@MrObvious. I've updated my comment. EXTERNAL is in between the 'auth' tags.

**Comment by fpqc** [comment]
@dorcas-ay He already said no on the quicker WSL updates.  It takes a lot of infrastructure to maintain separate update paths for each insider build.  They only release non-image updates for the slow and RC rings.  For fast ring it will always be a new image.

**Comment by therealkenc** [comment]
The "weird authentication thing" is issue #706. DBUS_COOKIE_SHA1 should work (ie !EXTERNAL), though I haven't tried because why. Per the `session.conf` [comments](https://apt-browse.org/browse/debian/jessie/main/amd64/dbus/1.8.20-0+deb8u1/file/etc/dbus-1/session.conf), the fallback exists for "obscure flavours of Unix that do not have a supported credentials-passing mechanism".

**Comment by dorcas-ay** [comment]
sorry! @MrObvious I've edited my comment above.

**Comment by MrObvious** [comment]
And that gets dbus working on Xenial for now?

**Comment by dorcas-ay** [comment]
Yea. That's what solved mine

**Comment by MrObvious** [comment]
`WARN  2016-09-07 09:47:43 unityfree <unknown>:0 In call to g_spawn_sync(), exit status of a child process was requested but ECHILD was received by waitpid(). Most likely the process is ignoring SIGCHLD, or some other thread is invoking waitpid() with a nonpositive first argument; either behavior can break applications that use g_spawn_sync either directly or indirectly.
WARN  2016-09-07 09:47:43 unityfree <unknown>:0 In call to g_spawn_sync(), exit status of a child process was requested but ECHILD was received by waitpid(). Most likely the process is ignoring SIGCHLD, or some other thread is invoking waitpid() with a nonpositive first argument; either behavior can break applications that use g_spawn_sync either directly or indirectly.
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error spawning command line 'dbus-launch --autolaunch=92b0ef7fbf17a91b9db8328956f30017 --binary-syntax --close-stderr': Child process killed by signal 97"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error spawning command line 'dbus-launch --autolaunch=92b0ef7fbf17a91b9db8328956f30017 --binary-syntax --close-stderr': Child process killed by signal 97"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error sending credentials: Error sending message: Invalid argument"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error sending credentials: Error sending message: Invalid argument"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error sending credentials: Error sending message: Invalid argument"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 4)
WARN  2016-09-07 09:47:43 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error sending credentials: Error sending message: Invalid argument"... Trying to reconnect (attempt 4)
^C
pstejska@PSTEJSKA-PC:~$ killaWARN  2016-09-07 09:47:46 unityfree <unknown>:0 In call to g_spawn_sync(), exit status of a child process was requested but ECHILD was received by waitpid(). Most likely the process is ignoring SIGCHLD, or some other thread is invoking waitpid() with a nonpositive first argument; either behavior can break applications that use g_spawn_sync either directly or indirectly.
lWARN  2016-09-07 09:47:46 unityfree <unknown>:0 In call to g_spawn_sync(), exit status of a child process was requested but ECHILD was received by waitpid(). Most likely the process is ignoring SIGCHLD, or some other thread is invoking waitpid() with a nonpositive first argument; either behavior can break applications that use g_spawn_sync either directly or indirectly.
lWARN  2016-09-07 09:47:46 unityfree <unknown>:0 In call to g_spawn_sync(), exit status of a child process was requested but ECHILD was received by waitpid(). Most likely the process is ignoring SIGCHLD, or some other thread is invoking waitpid() with a nonpositive first argument; either behavior can break applications that use g_spawn_sync either directly or indirectly.
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error spawning command line 'dbus-launch --autolaunch=92b0ef7fbf17a91b9db8328956f30017 --binary-syntax --close-stderr': Child process killed by signal 97"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error spawning command line 'dbus-launch --autolaunch=92b0ef7fbf17a91b9db8328956f30017 --binary-syntax --close-stderr': Child process killed by signal 97"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error spawning command line 'dbus-launch --autolaunch=92b0ef7fbf17a91b9db8328956f30017 --binary-syntax --close-stderr': Child process killed by signal 97"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Could not connect: No such file or directory"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error sending credentials: Error sending message: Invalid argument"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error sending credentials: Error sending message: Invalid argument"... Trying to reconnect (attempt 5)
WARN  2016-09-07 09:47:46 unity.glib.dbus.proxy GLibDBusProxy.cpp:196 Unable to connect to proxy: "Error sending credentials: Error sending message: Invalid argument"... Trying to reconnect (attempt 5)
 compiz`

Still not working.

**Comment by therealkenc** [comment]
Xenial looks pretty baked as of 14926. `do-release-upgrade` proceeds without any magic. The `SCM_CREDENTIALS` improvements make `gnome-terminal` happy over `AF_UNIX`. I ran into some Python3 related hell after upgrading from my overpackaged Trusty, but I very much doubt that's a WSL thing (I should have wiped and gone from scratch). YRMV, but it looks like we're good here now.

**Comment by netroby** [comment]
+1  please upgrade ubuntu to 16.04

**Comment by hachre** [comment]
This is being done, it's already out in the insider build 14951. 

More info here: https://t.co/p8HDaC6olx

**Comment by russalex** [comment]
As @hachre points out, we have updated from Trusty to Xenial on the Insider builds.  We will not be going back and updating the non-Insider build (14393) to Xenial.  There are some required fixes / improvements on the Insider track which we will not be bringing back to the Anniversary Update.

As always, I recommend jumping on the Insider track if you can.

**Comment by betam4x** [comment]
That's great. bit there are several big issues I personally have with the current versions of WSL.

1) You need an unstable Windows 10 build to get the latest bleeding edge updates...updates that typically fix the problems you are having.
2) Such a build is practically impossible to get (I've been enrolled in the fast ring for 2 months and haven't received a build beyond the anniversary version yet.)
3) If you do manage to obtain the builds, good luck!  From what I hear they are unstable (thanks to #2)
4)  My Macbook Pro is STILL a much better machine for development.  Keep in mind I'm not an Apple Fanboy, when I'm not working, I'm on my Windows 10 desktop.   However a Mac typically offers a better out of box experience overall.  A friendly UI, tabbed terminal, awesome open source support (homebrew, gcc, and rbenv come to mind), and best of all, it's not in beta.  It just works.

Don't get me wrong, I'd love to be able to use my Windows machine for development work...however at this point in time I only fire up my Windows box for 1 purpose...gaming...thanks to the above scenarios.  I use a Mac for everything else, and when I need something hybrid (that games and allows me to code) I use a laptop that dual boots Windows and Linux (Fedora to be exact).  WSL *ALMOST* changed that.  Almost.

**Comment by fpqc** [comment]
@betam4x Your insider setup is broken if it's taking a month.  Plus I have had no unstable rs2 builds thusfar.  Also, Rome wasn't built in a day, and it's like you're going into somebody else's half-built home and saying "tbh I like my living room better".  

Also, homebrew is much less convenient than apt or pacman and less powerful than something like pacaur.  Lastly, enjoy your unusable GUI on OSX.  At least if you were using straight Linux you could be forgiven by using Cinnamon or MATe. =p

**Comment by hachre** [comment]
@betam4x I agree that the dependance on Windows beta builds in order to get the latest WSL stuff is not ideal, but since the Windows kernel itself is tightly involved in achieving WSL, there is just no other option than that. Maybe one day when things are more mature there will be! I think you'll just have to wait a bit longer for things to progress and try to make the switch again in the future.

I definitely think things are heading in the right direction!

**Comment by scshepard** [comment]
Hi, this is mainly for "betam4x"; I'm super curious as to what you use your Mac for. I'm always looking for the best tools to get the job done, but am super happy with Windows 10, WSL; so, what am I missing??

Thanks!

________________________________

**Comment by betam4x** [comment]
OS-X uses a BSD variant as it's underpinnings so you get access to the full Unix stack with damn near perfect compatibility.  You also have better tools (simple example:  A 100% customizable tabbed terminal window), and things just work without issues.  I have used many languages/frameworks from C/C++ to high level stuff like Ruby and Java.  WSL has several issues when it comes to compatibility with many of the languages and frameworks I have used.  For example, with the release anniversary update, new Rails 5 apps won't work without tweaking the configs and removing the updated file watcher and disabling spring.  Apache SOLR also does not appear to work.  I've seen other incompatibilities elsewhere as well.

Without trying to turn this into a flame war, to @fpqc GUI is a matter of opinion.  I like the Mac UI, it works great, and things flow well.  I used to be a hardcore linux/windows user and hated the Mac UI, but once i actually learned it...my opinion changed.  It's easy to quickly get to where you need to be, and it just works.  Homebrew works great as well, it works just like apt, so i'm not sure what you meant by your comment.    Regarding your comment about Linux, I don't want to waste tons of time dealing with config issues, driver issues, etc.  On OS-X if i want to update to the next major version, I can easily do so without worrying if the system will boot afterwards.  I've run a couple distros of linux on my PC laptop (still have fedora installed), and all of them had issues installing (and ubuntu to this day does not work on the laptop).  Then they had issues with updating.  etc.  My time is valuable, soI just want a machine to work and work well.  Windows 10 with WSL comes close...but not quite.  Hopefully things will improve with WSL down the road...I really like being able to develop and game on the same machine.  Oh and Microsoft really needs a better terminal Window! (Yes I'm aware of Conemu, etc.)

**Comment by fpqc** [comment]
@betam4x All of these things are in the works (better terminal at #111 and Lx compatibility is improving every release.)

**Comment by aseering** [comment]
@betam4x -- as someone whose primary work machine is a Mac at the moment but who uses both Linux (multiple distributions) and Windows extensively as well:  I agree with all of your points.  (Though I have not found Apple to be as Linux-compatible as you have; BSD and Linux are similar but different enough that I tend to run into things that work on one and not the other periodically.)  I also hope that you'll consider re-evaluating WSL once it comes out of beta, or at least once it has matured a bunch; it sounds to me like it will be much more attractive to you at that point.

**Comment by russalex** [comment]
Wow, in meetings all day and just found this thread.  Interesting conversation.

@betam4x, thanks for your input. The main way we know what needs to improve is through feedback. I will not try and address everything you mention, but I can say that we are working hard to address a lot of the points you bring up. All I ask is that you give us another shot when the next release goes public (or before). It would be great to know if you're still having pain and if so what the bugs you're finding.

It does look like a lot of your issues are around the Insider program itself. I would be happy to get you in contact with someone on that team or with our customer support to help you with those issues if it helps.

**Comment by russalex** [comment]
One more point.  The original thread was upgrading Trusty to Xenial.  That bit of work is complete.  I believe this item can be closed.

**Comment by paulirwin** [comment]
@russalex Any chance we can get a technical blog post explaining what engineering changes it took for you guys to support Xenial in 14951? Just curious :-)

**Comment by fpqc** [comment]
@paulirwin They fixed Screen, sudo and a few other programs.  Apart from them, things were working since almost the first release.

**Comment by russalex** [comment]
@fpqc is pretty much correct.  In the early days we had some issues with the ps command as well, but overall Xenial has been working for some time.

Screen was mainly used for the upgrade process.  We did not want to swap over to Xenial until existing users could upgrade successfully.  The do-release-upgrade command relies on screen so we needed to get that working.

I would be happy to kick off a blog post on it if it was more of a push.  @paulirwin, anything specific that you're interested in here?  FYI: we do have a pty blog post on the back burner (haven't forgotten that one) and there are a couple of others that we are in the process of editing.

**Comment by DJviolin** [comment]
Can I ask an ETA when this Ubuntu 16.04 release will arrive in the stable branch for every users (I'm not want to upgrade my WIN to Insider Preview because of this)?

**Comment by fpqc** [comment]
@Djviolin If I remember correctly, it's planned for the end of March.  If I didn't remember correctly, the thing to search for is "Windows 10 Creative Update", which is the official name (and the planned release date was also part of the announcement of the official name, which is terrible.  Seriously, Microsoft Branding Guys, "Anniversary update" was already questionable, but I was willing to go with it because "oh, first anniversary, woohoo celebrate!!". But now "Creative update"? Come on. "Redstone II" is a far superior name.).

**Comment by whorfin** [comment]
I found this thread while repairing a xenial update, and thought I'd leave some notes here in the interests of helping the next person...

This was all done on multiple machines running Windows 10, Build 14393.693
So tl;dr protip - you don't need to be on insider to enjoy Xenial

First a note for Microsoft:  when remotely connecting to a WSL instance over `sshd/ssh`, the user is presented with the Canonical notice of upgrade availability, and the suggestion to run 
`do-release-upgrade`
So while this is hidden for those just using the shortcut, it shows up with ssh, and thus you'd [microsoft] be well advised to do something about educating us all...

The following worked for me, to upgrade a number of machines.
I chose not to hold back sudo, so be sure to have an alias somewhere or be in the habit:
`alias sudo='sudo -S'`

Removing `screen` allowed things to work fine interactively.
I use my own build of tmux [see #1063], so am independent of installed version, and it thus took me a while to understand that `screen` and `tmux` are secretly connected; removing one force-installs the other.  You do not want `screen` installed, or the upgrade will fail.

```
sudo -S apt-get remove screen
sudo -S apt-mark hold procps strace
sudo -S do-release-upgrade
```

Sometimes this works straightaway.
Sometimes it breaks, requiring a retry as follows
```
sudo -S dpkg --configure -a
sudo -S do-release-upgrade
```

Repeat until done and ready for next steps.
I had to answer a full-window prompt to "restart services"
On one system I had to review changes to `/etc/default/rcS`
On all systems so far I had to review a change which added the regular [non
-security] distro to `/etc/apt/apt.conf.d/50unattended-upgrades`
...and miscellaneous saying "y" to continue etc

After all that, the following:
```
sudo -S rm /etc/apt/apt.conf.d/50unattended-upgrades.ucftmp
sudo -S dpkg --configure -a
sudo -S apt-get update
sudo -S apt-get upgrade
sudo -S apt-get dist-upgrade
sudo -S apt-get autoremove
```

And good to go!

An aside - before I knew all this, I did the upgrade w/o the apt hold, and broke ps.
Recovering required not only the advice from others, but one more shared library package, as libprocps.so.3 was missing.
So, recovering a system with broken ps looks like this:
```
wget http://mirrors.kernel.org/ubuntu/pool/main/p/procps/procps_3.3.9-1ubuntu2_amd64.deb
wget http://mirrors.kernel.org/ubuntu/pool/main/s/strace/strace_4.8-1ubuntu5_amd64.deb
wget http://mirrors.kernel.org/ubuntu/pool/main/p/procps/libprocps3_3.3.9-1ubuntu2_amd64.deb
sudo -S dpkg -i procps_3.3.9-1ubuntu2_amd64.deb
sudo -S dpkg -i strace_4.8-1ubuntu5_amd64.deb
sudo -S dpkg -i libprocps3_3.3.9-1ubuntu2_amd64.deb
```


Cheers

**Comment by bugchecker** [comment]
@whorfin I tested your case without removing `screen` (https://github.com/Microsoft/BashOnWindows/issues/927#issuecomment-275846574). It works.

**Comment by whorfin** [comment]
I did not know about the `RELEASE_UPGRADER_NO_SCREEN` environment variable.
I will update my note

**Comment by bugchecker** [comment]
@whorfin Undocumented feature:
http://bazaar.launchpad.net/~ubuntu-core-dev/ubuntu-release-upgrader/trunk/view/head:/DistUpgrade/DistUpgradeMain.py#L222

**Comment by mpash** [comment]
Huge 🥂 to @whorfin for tutorializing the 16.04 upgrade for WSL. Works just fine now.

**Comment by DJviolin** [comment]
@whorfin Does your method have any advantage editing `/etc/apt/sources.list` directly and making the upgrade with these commands?

```sh
/usr/sbin/update-locale LANG=en_US.UTF8

echo "\
deb http://archive.ubuntu.com/ubuntu xenial main restricted universe multiverse\n\
deb http://archive.ubuntu.com/ubuntu xenial-updates main restricted universe multiverse\n\
deb http://archive.ubuntu.com/ubuntu xenial-backports main restricted universe multiverse\n\
deb http://security.ubuntu.com/ubuntu xenial-security main restricted universe multiverse" > /etc/apt/sources.list

apt update

apt -f install -y libc6
apt install -y --only-upgrade apt

rm -rf /var/lib/apt/lists/*
apt update
apt upgrade -y
apt full-upgrade -y
apt autoremove -y

rm /etc/apt/apt.conf.d/50unattended-upgrades.ucf-old
rm /etc/apt/apt.conf.d/50unattended-upgrades.ucf-dist
apt -f install
```

**Comment by bugchecker** [comment]
@DJviolin http://askubuntu.com/questions/409555/what-does-do-release-upgrade-really-do

**Comment by TechBossIceWobs** [comment]
OK, so I'll write up what worked for me, because I haven't seen this answer yet on this thread (but it is a long thread).

After installing, I opened my bash.
Did some apt-get upgrade, update, etc.
Sees that there is a new version available, runs do-release upgrade. Gets error and comes to this thread.

Finds https://github.com/Microsoft/BashOnWindows/issues/482#issuecomment-228089238. Thanks man! It does run, `sudo do-release-upgrade -f DistUpgradeViewNonInteractive -d # Screen is broken`. 

Oh, his advice fails because I get some `/proc/ error` and the workaround is unresponsive (no output after trying it three times). Grr..

Then I do `sudo -s` (it's my default for accessing the root user.)

And out of stupidity and desperation, I try `apt-get install`. I get the following output:

```
Reading package lists... Done
Building dependency tree
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 379 not upgraded.
```

379 not upgraded! Jackpot. Let's upgrade those! 

`apt-get upgrade` and y for yes. (Waits like 20 minutes.)

Waits even longer and then notices there was a yes/no something in the background. Says yes.

Test my new version with lsb_release -a. 
My output was the following:
```
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.2 LTS
Release:        16.04
Codename:       xenial
```

JACKPOT! And so that might be another way to upgrade for now.

P.S. I don't use Windows Insider or Fast Ring or anything, so this method can be used by anyone on the latest major version of Windows 10.

P.P.S. Noticed this particular warning prompt. Very interesting.
```
User defined time zone, leaving /etc/localtime unchanged.
Local time is now:      Wed Mar  1 17:12:15 STD 2017.
Universal Time is now:  Thu Mar  2 01:12:15 UTC 2017.
Run 'dpkg-reconfigure tzdata' if you wish to change it.
```
Images:

The original settings before starting my version of the workaround and after starting Manouchehri's workaround (thanks again man).
![image](https://cloud.githubusercontent.com/assets/7548448/23489974/65b3053c-feaa-11e6-8b5a-976ec9cdba65.png)

The beginning of my workaround (very simple!):
![image](https://cloud.githubusercontent.com/assets/7548448/23490010/8f7e791e-feaa-11e6-8d76-d55b2a58e9cd.png)

[I made the mistake of not recording an image of that purple screen with the yes and no prompt.]

The end. This is it. Finally upgraded.
![image](https://cloud.githubusercontent.com/assets/7548448/23490043/b6fc60f0-feaa-11e6-8ee2-a76adec87fab.png)

**Comment by NoSubstitute** [comment]
@whorfin Followed your instructions and upgraded to 16.04.2. Worked fine on 14393 AU.
Setting _alias sudo="sudo -S"_ afterwards is definitely the way to go.

**Comment by NoSubstitute** [comment]
So do I have to keep these held back, even after upgrading to Xenial, on 14393 AU?
**I'm not on Insider builds.**

The following packages have been kept back:
  procps strace

**Comment by NoSubstitute** [comment]
While running Xenial on 14393 must I keep the hold on procps and strace, lest they break my system if updated?

**Comment by NoSubstitute** [comment]
So is nobody following this thread anymore? (I know it's closed.)
Creators Update is coming soon for us non-insiders, so it's till interesting to know what will be needed when/after upgrading, and if this hold on procps/strace is permanent, or if they will be fine to update too.

**Comment by benhillis** [comment]
@NoSubstitute - I'm not sure if I understand the question, let me know if this doesn't answer it.

When you install Creator's Update you'll have two options to get to 16.04.
1. do-release-upgrade (without workarounds) to upgrade your existing WSL filesystem.
2. Reinstalling the Linux file system (lxrun /uninstall followed by lxrun /install)

Keeping specific packages at older versions is not a problem specific to WSL.  All of that logic lives in user-mode so any workarounds found for native Ubuntu should work.

**Comment by NoSubstitute** [comment]
Hi, @benhillis !
I already have xenial (16.04). I followed the instructions in here to get it on 14393 (not insider).
I was hoping I didn't have to re-install it after upgrading to the CU. And my second question was that, after I have installed CU, will I be able to let go of the "hold" on _procps_ and _strace_, since they haven't been updated since trusty.

**Comment by benhillis** [comment]
@NoSubstitute - If you've upgrade then you're good, you won't need to reinstall.  After you install CU you should be safe to remove the hold on procps and strace.  I'm running CU and the Xenial versions of those packages work for me.

**Comment by fpqc** [comment]
@benhillis By the way, do you know roughly when we'll start seeing the RS3 flights?

**Comment by jackchammons** [comment]
@fpqc No official word yet on when RS3 flighting will start. Historically, builds start to go out within a month after the previous version (Creators Update in this case) gets released to market.

**Comment by therealkenc** [comment]
For some value of historic...

14393 Insiders [July 18](https://blogs.windows.com/windowsexperience/2016/07/18/build14393/#Pr8VKXJGMerQ128i.97), and for the proletariat on [August 2](https://news.microsoft.com/2016/06/28/microsoft-announces-windows-10-anniversary-update-available-aug-2/#57H4ARHslJhqB3kR.97).
14901 [August 11](https://blogs.windows.com/windowsexperience/2016/08/11/announcing-windows-10-insider-preview-build-14901-for-pc/#fHe3jWmQODo4XPiy.97).

It was 15 days from Insiders to RTM, and 9 days from RTM to the next Insiders. The street says RS2 on April 11. YRMV.

**Comment by fpqc** [comment]
@therealkenc Yeah but I think they're going to be a bit more cautious this time given some of the bad press around RS1 (webcams and such).

**Comment by Scandalon** [comment]
Thanks @TechBossIceWobs , that worked for me.

**Comment by Scandalon** [comment]
Just a reminder to everyone that upgraded WSL early, as the "creators update" is installed, to not forget to "apt-mark unhold procps strace sudo". (and probably a dist-upgrade too.)

**Comment by fullerth** [comment]
For those of you still stuck on old windows builds who didn't see the tutorial from @whorfin before upgrading, the kernel.org mirror's have moved. Here's the current commands to fix your `sudo`/`ps`/`strace`:

```
wget http://mirrors.edge.kernel.org/ubuntu/pool/main/s/sudo/sudo_1.8.9p5-1ubuntu1.4_amd64.deb
wget http://mirrors.edge.kernel.org/ubuntu/pool/main/p/procps/procps_3.3.9-1ubuntu2_amd64.deb
wget http://mirrors.edge.kernel.org/ubuntu/pool/main/s/strace/strace_4.8-1ubuntu5_amd64.deb
wget http://mirrors.edge.kernel.org/ubuntu/pool/main/p/procps/libprocps3_3.3.9-1ubuntu2_amd64.deb
dpkg -i sudo_1.8.9p5-1ubuntu1.4_amd64.deb
dpkg -i libprocps3_3.3.9-1ubuntu2_amd64.deb
dpkg -i procps_3.3.9-1ubuntu2_amd64.deb
dpkg -i strace_4.8-1ubuntu5_amd64.deb
```
